src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel_UpdateViewModelState.cs

```diff
@@ -0,0 +1,151 @@
+        private ExpansionOption? _currentExpansionOption = null;
+
+        private async ValueTask UpdateViewModelStateAsync(ImmutableSegmentedList<ViewModelStateData> viewModelStateData, CancellationToken cancellationToken)
+        {
+            var model = await _documentSymbolQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
+            Assumes.NotNull(model); // This can only be null if no work was ever enqueued, we expect at least one item to always have been queued in the constructor
+
+            var searchText = viewModelStateData.SelectLastOrDefault(static x => x.SearchText);
+            var position = viewModelStateData.SelectLastOrDefault(static x => x.CaretPositionOfNodeToSelect);
+            var expansion = viewModelStateData.SelectLastOrDefault(static x => x.ExpansionOption);
+            var dataUpdated = viewModelStateData.Any(static x => x.DataUpdated);
+
+            if (searchText is not null)
+            {
+                _currentSearchText = searchText;
+            }
+
+            if (position is not null)
+            {
+                _currentCaretPosition = position;
+            }
+
+            // These updates always require a valid model to perform
+            if (!model.IsEmpty)
+            {
+                if (_currentSearchText is null && dataUpdated)
+                {
+                    var documentSymbolViewModelItems = DocumentOutlineHelper.GetDocumentSymbolItemViewModels(model.DocumentSymbolData);
+                    ApplyExpansionStateToNewItems(documentSymbolViewModelItems, DocumentSymbolViewModelItems);
+                    DocumentSymbolViewModelItems = documentSymbolViewModelItems;
+                }
+
+                if (_currentSearchText is { } currentQuery)
+                {
+                    ImmutableArray<DocumentSymbolDataViewModel> documentSymbolViewModelItems;
+                    if (currentQuery == string.Empty)
+                    {
+                        // search was cleared, show all data.
+                        documentSymbolViewModelItems = DocumentOutlineHelper.GetDocumentSymbolItemViewModels(model.DocumentSymbolData);
+                        ApplyExpansionStateToNewItems(documentSymbolViewModelItems, DocumentSymbolViewModelItems);
+                    }
+                    else
+                    {
+                        // We are going to show results so we unset any expand / collapse state
+                        _currentExpansionOption = null;
+                        documentSymbolViewModelItems = DocumentOutlineHelper.GetDocumentSymbolItemViewModels(
+                             DocumentOutlineHelper.SearchDocumentSymbolData(model.DocumentSymbolData, currentQuery, cancellationToken));
+                    }
+
+                    DocumentSymbolViewModelItems = documentSymbolViewModelItems;
+                }
+
+                if (_currentCaretPosition is { } currentPosition)
```

Comment:
Can we ever have two UpdateViewModelStateAsync running at the same time here? If we can, what is ensuring ordering that the fields are being assigned the final values in the first place? Something still seems very sketchy here in terms of threading guarantees.

Reply:
`AsyncBatchingWorkQueue` is intended to ensure that `UpdateViewModelStateAsync` does not have multiple instances running concurrently.